core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -83,28 +100,31 @@ public static void removeTypes() {
         typesContext.remove();
     }
 
-    private final Version indexVersionCreated;
-
-    private final IndexQueryParserService indexQueryParser;
-
     private final Map<String, Query> namedQueries = new HashMap<>();
-
     private final MapperQueryParser queryParser = new MapperQueryParser(this);
-
+    private final IndicesQueriesRegistry indicesQueriesRegistry;
     private boolean allowUnmappedFields;
-
     private boolean mapUnmappedFieldAsString;
-
     private NestedScope nestedScope;
-
     private QueryParseContext parseContext;
+    boolean isFilter; // pkg private for testing
 
-    boolean isFilter;
+    public QueryShardContext(IndexSettings indexSettings, Client client, BitsetFilterCache bitsetFilterCache, IndexFieldDataService indexFieldDataService, MapperService mapperService, SimilarityService similarityService, ScriptService scriptService,
+                             final IndicesQueriesRegistry indicesQueriesRegistry) {
+        this.indexSettings = indexSettings;
+        this.scriptService = scriptService;
+        this.client = client;
+        this.similarityService = similarityService;
+        this.mapperService = mapperService;
+        this.bitsetFilterCache = bitsetFilterCache;
+        this.indexFieldDataService = indexFieldDataService;
+        this.allowUnmappedFields = indexSettings.isDefaultAllowUnmappedFields();
+        this.indicesQueriesRegistry = indicesQueriesRegistry;
+        this.parseContext = new QueryParseContext(indicesQueriesRegistry);
+    }
 
-    public QueryShardContext(IndexQueryParserService indexQueryParser) {
-        this.indexVersionCreated = indexQueryParser.getIndexCreatedVersion();
-        this.indexQueryParser = indexQueryParser;
-        this.parseContext = new QueryParseContext(indexQueryParser.indicesQueriesRegistry());
+    public QueryShardContext clone() {
+        return new QueryShardContext(indexSettings, client, bitsetFilterCache, indexFieldDataService, mapperService, similarityService, scriptService, indicesQueriesRegistry);
```

Comment:
ops..... I see now the clone impl... looks good :) maybe do it through a copy constructor would not make people think (like I did) that we call Object#clone... minor thing though